### PR TITLE
[Converged MME] Update location procedure flags addition.

### PR DESCRIPTION
**Pull request by Vyas20**
Signed-off-by: Mohit Vyas <mohit.vyas@openairinterface.org>

[Converged MME] Update location procedure flags addition

## Summary

Adding flags (Dual registration 5g indicator flag ) in update location request message.

 ## Test Plan

S1AP Tester, dsTester, Commercial handset (Oppo) and Quectel modules.

### CODE REVIEW

**Review by themarwhal — APPROVED** [review]
feg changes lgtm :)

File: feg/protos/s6a_proxy.proto
Comment (themarwhal) [inline feg/protos/s6a_proxy.proto]: minor nit but spacing looks a bit off here.

**Comment by ulaskozat** [comment]
@Vyas20 there are a number of github actions warnings, please comment if these are legit or not.

File: feg/protos/s6a_proxy.proto
Comment (uri200) [reply feg/protos/s6a_proxy.proto]: We normally don't abbreviate on protos. Let's use the full name. 

Reason is that protons are used across different services and its field name can help to provide context. 

```
bool send_dual_registration_5g_indicator
```

**Review by uri200 — CHANGES_REQUESTED** [review]

File: lte/gateway/c/oai/lib/s6a_proxy/itti_msg_to_proto_msg.cpp
Comment (ssanadhya) [inline lte/gateway/c/oai/lib/s6a_proxy/itti_msg_to_proto_msg.cpp:103]: @Vyas20 , this PR does not include how the `dual_regis_5g_ind` value is set in the `msg`. It needs to be set by MME here https://github.com/magma/magma/blob/master/lte/gateway/c/oai/tasks/nas/emm/Authentication.c#L1452 based on the flags seen in the Attach Request message

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/s6a/CMakeLists.txt]: Why do we need TASK_MME_APP here?

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/s6a/CMakeLists.txt]: Why do we need TASK_MME_APP here?

File: feg/protos/s6a_proxy.proto
Comment (ssanadhya) [inline feg/protos/s6a_proxy.proto]: Where is this being used?

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: btw, can you correct indentation of those comments please

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/s6a/CMakeLists.txt:34]: I added mme_app_ue_conetxt.h which had dependency on  mme_app_sgs_fsm.h so added this task here.

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/s6a/CMakeLists.txt:24]: I added mme_app_ue_conetxt.h in s6a_auth_info.c which had dependency on  mme_app_sgs_fsm.h so added this task here.

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: Per our discussion here https://github.com/magma/magma/issues/5672

shouldn't we use `Feature-List-ID-2 {NR as Secondary RAT}`

**Review by uri200 — CHANGES_REQUESTED** [review]

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: I don't understand what you are trying to achieve here. You are allocating a memory of size of pointer (8 bytes), you are accessing then an arbitrary memory location. Were you intending to get ue_context_p from hash table via ue id and use it instead?

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: I don't understand why you malloced this data structure in the first place.

File: feg/protos/s6a_proxy.proto
Comment (uri200) [reply feg/protos/s6a_proxy.proto]: Since potentially we can have more flags enabled from  `Feature-List-ID-2`  in the future, lets create a message for itself. Changing protos is very difficult once they are in place

```

message AuthenticationInformationRequest {
...
FeatureListId2 feature_list_id_2 = 8;
...
}

message FeatureListId2 {
	bool nr_as_secondary_rat = 1;
}

```

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: isn't it this a bad field?

File: feg/protos/s6a_proxy.proto
Comment (uri200) [reply feg/protos/s6a_proxy.proto]: per 
http://www.eventhelix.com/5G/non-standalone-access-en-dc/en-dc-secondary-node-addition-core-network-details.pdf
I see it uses `Update Location Request`, but I don't see `AuthenticationInformationRequest` Is it really necessary to include that field on this message? Maybe it is only necessary on UpdateLocationReques/Response?

**Review by uri200 — CHANGES_REQUESTED** [review]

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/s6a/CMakeLists.txt]: Removed

File: lte/gateway/c/oai/tasks/s6a/CMakeLists.txt
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/s6a/CMakeLists.txt]: Removed

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: Resolved

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: Resolved

File: feg/protos/s6a_proxy.proto
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: I am sorry I didn't understand , Can you elaborate ?

File: feg/protos/s6a_proxy.proto
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: According to 3GPP 29.272 V-15.10.0, Supported features(as an Optional field) AVP should be added in Authentication Information Request. NR as secondary RAT is one of the bit field in Feature list ID2 which is a part of Supported features AVP.

**Comment by rdefosse** [comment]
@Vyas20 @ulaskozat 

Last run w/ DsTester failed any attach. Mohit will have to debug in this env.

Raphael

File: feg/protos/s6a_proxy.proto
Comment (uri200) [reply feg/protos/s6a_proxy.proto]: Is `dual_registration_5g_indicator` necessary?
I know we discussed this and I may have forgotten something. But is this field related with the feature of adding a secondary rat `nr_as_secondary_rat`

The protocol talks about a deregistration. Per that description I dont see the relationship with nr_as_secondary_rat

image

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/oai/lib/s6a_proxy/itti_msg_to_proto_msg.cpp
Comment (ulaskozat) [reply lte/gateway/c/oai/lib/s6a_proxy/itti_msg_to_proto_msg.cpp:103]: @Vyas20 can you address/respond to @ssanadhya's comment?

Comment (Vyas20) [reply lte/gateway/c/oai/lib/s6a_proxy/itti_msg_to_proto_msg.cpp:103]: Added flag nr_as_secondary_rat in attach request message as per  3GPP 29.272 V-15.10.0.

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: lets use `BitrateUnit`

Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: Changed

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: remove one space

File: feg/protos/s6a_proxy.proto
Comment (uri200) [inline feg/protos/s6a_proxy.proto]: remove space

**Comment by uri200** [comment]
note I add support for this at feg on https://github.com/magma/magma/pull/6234

File: feg/protos/s6a_proxy.proto
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: Updated

File: feg/protos/s6a_proxy.proto
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: Updated

File: lte/gateway/c/oai/common/common_types.h
Comment (ulaskozat) [inline lte/gateway/c/oai/common/common_types.h:201]: how is this used?

File: lte/gateway/c/oai/common/common_types.h
Comment (ulaskozat) [inline lte/gateway/c/oai/common/common_types.h]: is this supposed to be boolean?

File: lte/gateway/c/oai/common/common_types.h
Comment (Vyas20) [reply lte/gateway/c/oai/common/common_types.h:201]: not yet used, I added for future usage if we require more features.

File: lte/gateway/c/oai/common/common_types.h
Comment (Vyas20) [reply lte/gateway/c/oai/common/common_types.h]: Changed to boolean.

**Review by ulaskozat — APPROVED** [review]
LGTM

File: feg/protos/s6a_proxy.proto
Comment (ulaskozat) [inline feg/protos/s6a_proxy.proto]: make BPS as 0 as the default value...

**Review by uri200 — CHANGES_REQUESTED** [review]
Please rebase since proto files have been changed to include your changes to support your fields in FEG https://github.com/magma/magma/pull/6234

**Review by uri200 — APPROVED** [review]

File: feg/protos/s6a_proxy.proto
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: Done

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (ardzoht) [inline lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: this is missing the ue_id param to be logged

**Review by ardzoht — APPROVED** [review]
one minor comment on the OAILOG_DEBUG usage (missing `ue_id` param), overall lgtm

File: lte/gateway/c/oai/tasks/nas/emm/Authentication.c
Comment (Vyas20) [reply lte/gateway/c/oai/tasks/nas/emm/Authentication.c]: I had used this log as a debug msg which is now removed.

File: lte/gateway/python/magma/subscriberdb/protocols/s6a_proxy_servicer.py
Comment (ulaskozat) [inline lte/gateway/python/magma/subscriberdb/protocols/s6a_proxy_servicer.py]: line is too long.

File: lte/gateway/c/oai/common/state_converter.cpp
Comment (ulaskozat) [inline lte/gateway/c/oai/common/state_converter.cpp]: why is this not read from state_ambr?

File: lte/gateway/c/oai/tasks/nas/nas_state_converter.cpp
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/nas/nas_state_converter.cpp]: why is this not get from state_ambr.bitsunit?

File: lte/protos/oai/common_types.proto
Comment (ulaskozat) [inline lte/protos/oai/common_types.proto]: why is the order reverse here?

File: lte/gateway/c/oai/common/common_types.h
Comment (ulaskozat) [inline lte/gateway/c/oai/common/common_types.h]: make field name bitsunit consistent with the proto definition i.e., `bitrate_units`.

File: lte/gateway/c/oai/common/common_types.h
Comment (ulaskozat) [inline lte/gateway/c/oai/common/common_types.h]: `bitrate` rather than `bits`

File: lte/gateway/c/oai/common/common_types.h
Comment (ulaskozat) [inline lte/gateway/c/oai/common/common_types.h]: let's make this naming consistent with `br_ul` and `br_dl`, i.e., `br_unit`.

**Comment by rdefosse** [comment]
@Vyas20 @ulaskozat @lionelgo 

The run on the OAI pipeline was a false negative.

We upgraded dsTester to `5.6` and some script syntax to detect the end of test is now incorrect. I've just fixed it.

Only this PR and another one were impacted.

Raphael

File: lte/protos/oai/common_types.proto
Comment (Vyas20) [reply lte/protos/oai/common_types.proto]: Resolved

File: lte/gateway/c/oai/common/common_types.h
Comment (Vyas20) [reply lte/gateway/c/oai/common/common_types.h]: Resolved

File: lte/gateway/c/oai/common/common_types.h
Comment (Vyas20) [reply lte/gateway/c/oai/common/common_types.h]: Resolved

File: lte/gateway/c/oai/common/common_types.h
Comment (Vyas20) [reply lte/gateway/c/oai/common/common_types.h]: Resolved

**Comment by ekowtaylor** [comment]
Once this PR has been approved and passing all CI checks, use the ready2merge label to indicate the maintainers can merge your PR.

**Pull request merged**
